### DISCUSSION TITLE
promises, suspense, and changing params

### DISCUSSION THREAD

**Discussion by joprice** [Omnistac/zedux#132] [Q&A] [answered]
When using an atom that wraps a promise along with suspense, Is is possible to trigger suspension when props passed to the atom as args change? Or is this a case where `suspend: false` should be used? In other words, is suspense only triggered on the first load (and an invalidation?) of any atom. 

I saw a relevant comment here https://github.com/Omnistac/zedux/blob/c1e075ac337ea8359f1a600a50b5088a7572a254/packages/react/test/snippets/suspense2.mdx#L40 `React transitions don't apply to suspense. But if they ever do, this will be cool` which seems to imply that I should use my own loading code making use of `PromiseState` when I care about subsequent re-loads showing the loader.

**Comment by bowheart** [comment, marked as answer]
Hey @joprice. Components using any atom hooks will re-suspend when the atom's `.promise` reference changes.

```ts
const suspendingAtom = atom('suspending', () => {
  // quick example that changes the promise reference every time this atom reevaluates:
  return api().setPromise(new Promise(resolve => setTimeout(() => resolve('test'), 1000))))
})

function ExampleComponent() {
  const suspendingInstance = useAtomInstance(suspendingAtom)

  return <button onClick={() => suspendingInstance.invalidate()}>Re-suspend</button>
}

function ExampleSuspenseWrapper() {
  return (
    <Suspense fallback={<div>ExampleComponent suspending...</div>}>
      <ExampleComponent />
    <Suspense>
  )
}
```

Note that you can use `injectMemo` or `injectPromise` to prevent the promise reference from changing every evaluation, which is usually what you want.

>  Is it possible to trigger suspension when props passed to the atom as args change?

This should also "just work" to re-suspend the component, but in this case, it'll be because the different params create an entirely new atom instance.

```ts
function ExampleComponent() {
  const [counter, setCounter] = useState(0)
  // creates a new atom instance every time `counter` changes.
  const suspendingInstance = useAtomInstance(suspendingAtom, [counter])

  return <button onClick={() => setCounter(state => state + 1)}>Re-suspend</button>
}
```

(an example like this should be paired with `ttl` to prevent creating tons of dead atom instances).

Transitions are completely different from suspense. It doesn't sound like you're referring to transitions (via `useTransition`) here (tell me if I'm wrong). Transitions are a pretty half-baked React feature right now and are especially poor for 3rd-party state managers, though Zedux works around some of the difficulties to provide at least the best possible transition support. But again, I don't think that's what you're looking for.

**Reply by joprice** [reply]
Yea this all makes sense now. I have been using `injectPromise`, mostly for the convenience of getting an `AbortController` instance. 

The main use case I have right now is to have an `ErrorBoundary` wrapping `Suspense`, where calling `resetErrorBoundary` would refresh the data after a failure. It currently caches it, as is the intended behavior you mentioned above, so I resorted to calling `.destroy(true)` before `resetErrorBoundary`.

In order to mimic the above pattern of swapping the promise reference, I guess I would write my own injector that basically inlines the behavior of `injectPromise` with `dataOnly`? Or perhaps add a dummy dependency of a ref that increments on each render and pass that to `injectPromise`?

**Reply by joprice** [reply]
I just noticed `runOnInvalidate`, which seems to have baked in already what I was suggesting to try.

**Reply by joprice** [reply]
`runOnInvalidate` fixes reload without requiring `destroy(true)`. 

For "transitioning", where I basically want to keep the previous value and show a loading overlay, I ended up putting the parameters of the atom into another atom using the set once idea from a previous discussion, where the last non-null value is kept, along with ttl of 0.  I then store the previous value in that atom using a ref. This allows a modal to show a loader when first displayed, then an overlay on subsequent loads, and automatically clear it's value on close so the previous value does not display when reopening the modal.  Since the params are only referenced inside the second atom, the outer one acts as a slot to allow swapping a single value, even though a new atom is being created internally per param. This works great and I can see using it in any similar situation. Thanks for all the tips!

**Reply by bowheart** [reply]
Awesome! Yep, `runOnInvalidate` was what I was going to suggest.

Sorry I was slow responding last week. I'm back to normal this week.

**Reply by joprice** [reply]
I'm hitting one more broad issue with error handling: When a transient error like network a blip that bubbles up to an `ErrorBoundary` higher up in the tree, outside of the context where the original atom that caused it, it's not clear what is meant to be invalidated to get it to retry, as there is not necessarily an association with the error and the atom. Calling `resetErrorBoundary` will trigger a re-render, but end up reusing the failed promise, so no new http call is made. `ecosystem.reset` can be used to reset everything, but if there are expensive data loads that happen on app start, that causes unnecessary overhead for retrying a single transient error. 

A few ideas I had:
* use a custom error subclass to attach a field that has the atom or an id attached, and then walk the error's cause chain, in case the exception was rethrown. 
* use the ecosystem to filter for atom's in failed states (using PromiseState) to invalidate them
* wrap the access to the atom in a catch block that invalidates and rethrows the error, so it's auto-invalidated on failure. I'm not sure this will account for transitive invalidation, if an atom depends on another that was the root cause, does that inner one need to be invalidated? 

Any thoughts on how to approach this?

**Reply by joprice** [reply]
I tried out the second one and it works, currently checking for atom's with non-null `_promiseError` and invalidating them. I guess I could also check for Active lifecycle status as well but not sure the implications of that.

**Reply by bowheart** [reply]
The first bullet was my first thought. I don't think I see a problem with it.

The second bullet is a cool idea. I'm not sure if I would have thought of that. It's really good if you don't mind globally searching all atoms. You could filter for specific ones that are "globally invalidatable" e.g. with specific atom key prefixes or using [atom flags](https://omnistac.github.io/zedux/docs/walkthrough/configuring-atoms#flags) to narrow that search.

I think I have a general suggestion for you. Atoms in Zedux are very autonomous. They're great at handling operations by themselves, without requiring consumers to do anything imperative. My general approach thinking about this problem is, "Okay, this atom threw an error. How can he set himself up to retry the operation?"

Whether that retry is automatic or happens on button click or some other event, the atom that errored should be setting up the retry, not any of its transitive dependents or any React components.

So following that line of thinking, I have this (roughly coded) solution:

```ts
const retriesAtom = atom('retries', () => {
  const store = injectStore<(() => any)[]>([]) // a list of functions to run on retry

  const addRetry = (fn: () => any) => store.setState(state => [...state, fn])

  const retryAll = () => {
    for (const fn of store.getState()) {
      fn()
    }

    store.setState([])
  }

  return api(store).setExports({ addRetry, retryAll })
})

const maybeFailingAtom = ion('maybeFailing', ({ getInstance }) => {
  const self = injectSelf()

  const promise = injectPromise(
    () => maybeFetchSomething().catch(
      // this atom sets itself up for retry:
      () => getInstance(retriesAtom).exports.addRetry(() => self.invalidate())
    ),
    [],
    { runOnInvalidate: true }
  )

  return api(promise)
})

function ErrorBoundaryWrapper() {
  const retriesInstance = useAtomInstance(retriesAtom)

  return (
    <ErrorBoundary
      FallbackComponent={MyFallbackComponent}
      onReset={() => {
        retriesInstance.exports.retryAll()
      }}
    >
      <ComponentSubtreeThatUsesMaybeFailingAtom />
    </ErrorBoundary>
  )
}
```

Hopefully that helps with approaching problems like this in Zedux. Your first bullet pretty much does this, just having the atom throw a particular, smart error instead of using a separate `retriesAtom`. Your second bullet does too, actually, discreetly flagging atoms as "retryable" based on their `_promiseStatus` (or `_promiseError`, sure), though, again, I feel like that's too global and could end up flagging atoms that it shouldn't.

Watch out for underscore-prefixed properties btw. Lots of those will change in v2. I'll always try to show how to update all of them in version migration notes, so watch for those. These `_promise*` properties will probably change.

**Reply by joprice** [reply]
This is a great suggestion. It is less global, opt in, and explicit. Plus, I was already wary about relying on the internals with the promise state. I'll prototype this in my codebase and see If I hit any issues with it. Thanks again!

**Reply by joprice** [reply]
I finally ended up trying this idea out, as I found that searching for atoms is a overkill, causing unnecessary invalidations. I could use the graph to narrow it down, following dependencies and searching for the root cause, but I have a feeling it's going to lead to other surprises, and retrying things I didn't intend to be retried like non-idempotent operations.

When trying out the retry atom approach, I hit one issue with your example code in the context of an atom that has a dynamic edge via, for instance, `injectAtomValue`. Such an atom is effectively parameterized by another atom, "internally" or "implicitly" instead of explicitly via params. If the root cause of the error is an issue with a particular value of the params, which is often the case for my current data, and changing that value via some of button in the app resolves the issue, the retry should be removed from the list, as it's no longer required since the data was refreshed already. This also allows automatically dismissing an error boundary if the data has been refreshed and the retries atom is now empty.

```ts
 export const retriesAtom = atom('retries', () => {
    const store = injectStore<Record<string, () => any>>({});

    const addRetry = (id: string, fn: () => any) => store.setState(state => ({...state, [id]: fn}));
    const removeRetry = (id: string) =>
      store.setState(state =>
        Object.fromEntries(Object.entries(state).filter(([key, _value]) => key !== id)),
      );

    const retryAll = () => {
      for (const [_, fn] of Object.entries(store.getState())) {
        fn();
      }
      store.setState({});
    };

    return api(store).setExports({addRetry, removeRetry, retryAll});
  });
```

The data is keyed by `self.id` and then the entry is removed in the `then` of the promise with `getters.getInstance(retriesAtom).exports.removeRetry(self.id);`. I haven't yet depended on the `id` of the atom anywhere, so I was curious if it's safe to make use of that for such a use case or if you see any other problems with the extended approach.

**Reply by bowheart** [reply]
Beautiful, yes a `removeRetry` export like that would be the first thing I'd add to my example. It may also need a `clearAll`.

Lazily reading an atom instance's `id` will always be safe. The only case to watch out for with ids (which shouldn't be a problem here) is with scoped atoms in Zedux v2. This case:

```ts
const { id } = injectSelf() // snapshots the current value
inject(someContext) // add scope! Changes the atom's id after initial evaluation
```

I'll make sure this is documented. As long as you

- don't snapshot the id, potentially closing over a stale value
- access the id after initial evaluation, like in your `promise.then`

the id is safe to use. This case is only something to watch out for when using `injectSelf` specifically - consumers of the atom can always use `atomInstance.id` safely; the id becomes stable after initial evaluation, before any consumers can access it.
